File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/content/PseudoTextBridge.swift
//
//  PseudoTextBridge.swift
//  StyleEngine/content — wave 42, lane W2 (the iOS pseudos seam).
//
//  The PURE half of the ordinary-element generated-content path: one
//  `pseudos.<role>` bucket (the extractor-owned payload the wire forwards
//  VERBATIM, schema/spec/01-envelope.md `pseudos` row) → the inline TEXT
//  that bucket bakes, or nil when the bucket carries nothing this bridge
//  can honestly fold. No SwiftUI types, so every rule is XCTest-pinnable
//  without a simulator — the same split RootPseudoSpec uses for the
//  body-root box family.
//
//  ── WHY THIS EXISTS ──────────────────────────────────────────────────
//  v2 IR components carry baked ::before/::after generated content:
//  `pseudos.before._text` is the extractor's full css-content-3 §2 /
//  css-lists-3 §5 resolution (`content: "B" counter(foo)` → "B7", tagged
//  `_lossyReasons: ["generated-content-baked"]`). Web renders it as an
//  inline <span> (runtimes/web/src/renderer/PseudoNodeRenderer.ts); iOS
//  rendered NOTHING for every non-body-root component — MEASURED in
//  tools/titan/runs/wave41-final/sections/css-lists: the Chromium ref for
//  counter-reset-reversed-pseudo-001 shows "B7A5"/"B4A2" at the top-left,
//  the iOS capture is completely blank (ios-screenshots/…pseudo-001.png).
//
//  ── WHY TEXT-ONLY, AND STRICTLY GATED ────────────────────────────────
//  The natives cannot hand a raw declarations map to a DOM like web can.
//  The corpus splits pseudo buckets into two populations: TEXT pseudos
//  (content + already-baked counter-* — the css-lists/css-counter-styles
//  fail set) and BOX pseudos (display/width/height/background/position —
//  the css-contain and css-anchor-position families). This bridge folds
//  ONLY the first: a bucket declaring anything beyond the consumable set
//  is refused whole (never half-rendered) and every refused declaration
//  is named in the log, per the repo's no-silent-fallthrough rule.
//  The body-root BOX population keeps its dedicated RootPseudoBox path.
//

import Foundation

/// Pure bridge: one `pseudos.<role>` bucket → the baked inline text.
// enum: pure namespace — never instantiated (mirrors RootPseudo et al.).
enum PseudoTextBridge {

    /// Declarations a text fold can CONSUME or safely treat as inert:
    /// `content` is the very source of the baked `_text` (css-content-3
    /// §2), and the counter machinery (`counter-increment` / `counter-reset`
    /// / `counter-set`, css-lists-3 §4) is already RESOLVED into that text
    /// by the extractor's bake — re-running it here could only disagree.
    static let inertDeclarations: Set<String> = [
        "content", "counter-increment", "counter-reset", "counter-set",
    ]

    /// The inline text this bucket bakes, or nil when there is nothing an
    /// inline text run can honestly express (empty content, an unbaked
    /// function value, or any declaration outside the consumable set).
    ///
    /// - Parameters:
    ///   - bucket: the wire object — `{ "properties": {…}, "_text": "…" }`,
    ///     the shape the WPT extractor emits and the decoder forwards
    ///     verbatim (spec 01: `pseudos` payloads are never flattened).
    ///   - role: "before" / "after", for the log lines only.
    static func inlineText(bucket: IRValue?, role: String) -> String? {
        // A bucket is an object; anything else is not a pseudo payload.
        guard let b = bucket, b.objectValue != nil else { return nil }
        // The raw CSS declarations map — extractor-owned, spec 01. Absent
        // map = a bare text bucket; tolerated (nothing to gate on).
        let decls = b["properties"]?.objectValue ?? [:]
        // Gate pass: name EVERY declaration outside the consumable set,
        // then refuse the whole bucket — a pseudo asking for a box
        // (display/width/background) or per-run styling (color/font-*)
        // must not be silently flattened into the host's uniform text run.
        var refused = false
        // Sorted so multi-key refusals log in a deterministic order.
        for prop in decls.keys.sorted() {
            // Normalise the key the way RootPseudoSpec does: CSS property
            // names are ASCII case-insensitive (css-syntax-3 §5).
            let key = prop.trimmingCharacters(in: .whitespaces).lowercased()
            // Consumable/inert → no objection from this declaration.
            guard !inertDeclarations.contains(key) else { continue }
            // No silent fallthrough: the refused declaration is named once
            // per (role, property) — not per component, so a corpus-wide
            // repeat cannot flood the capture log.
            PropertyTracker.logOnce(
                key: "pseudotext-unsupported-\(role)-\(key)",
                message: "[PseudoText] ::\(role) declaration '\(prop)' is beyond the inline-text fold — bucket not folded")
            // Remember the refusal but keep scanning so every offending
            // declaration gets its own named log line.
            refused = true
        }
        // Any refusal keeps the bucket unrendered — exactly the behaviour
        // this platform had before the seam, now named instead of silent.
        guard !refused else { return nil }
        // The baked literal: `_text` is what the extractor emits today;
        // `text` is the v2 spelling — web tolerates both in exactly this
        // order (PseudoNodeRenderer.ts `p._text ?? p.text`), so mirror it.
        guard let text = b["_text"]?.stringValue ?? b["text"]?.stringValue else {
            // A `content` declaration with NO baked text means the extractor
            // could not resolve it (e.g. `content: counter(foo)` on a
            // display:none subtree that was never laid out) — web renders
            // the empty string for these too; name the gap and fold nothing.
            if decls.keys.contains(where: { $0.trimmingCharacters(in: .whitespaces).lowercased() == "content" }) {
                PropertyTracker.logOnce(
                    key: "pseudotext-unbaked-\(role)",
                    message: "[PseudoText] ::\(role) content has no baked text (_text absent) — rendered as empty")
            }
            return nil
        }
        // `content: ""` bakes an EMPTY string — a legal pseudo that paints
        // no glyphs (css-content-3 §2), so there is nothing to fold.
        return text.isEmpty ? nil : text
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/content/PseudoTextBridge.kt
package com.styleconverter.runtime.content

// PseudoTextBridge — wave-50 lane B3: the PURE half of the Compose ::after
// text fold, and the byte-parallel twin of iOS
// StyleEngine/content/PseudoTextBridge.swift (same name, same gate set, same
// refusal vocabulary; skeptic probes diff the two).
//
// The split mirrors the iOS pair exactly: [PseudoTextFold] owns the COMPONENT
// rewrite (which bucket may be claimed, and where the glyphs land), this file
// owns the BUCKET question — can an inline text run honestly express this
// `pseudos.after` payload at all? Keeping it separate also keeps both files
// inside the repo's 200-line target and lets the JVM pin table exercise the
// declaration gate without building a component.
//
// WHAT AN INLINE RUN CAN CARRY. The bucket's `properties` map is RAW CSS
// strings (schema/spec/01-envelope.md forwards the extractor payload
// verbatim) and this runtime has no CSS parser, so the gate is a whitelist,
// not a parse: declarations that are inert for a text run pass, the three
// styling declarations PseudoStyledDeclarations can type pass as typed
// properties, and EVERYTHING else refuses the whole bucket — named through
// the PropertyTracker, never silently flattened. A pseudo that asked for a
// 20x20 tomato square (css-cascade/scope-pseudo-element) must not be folded
// down to the letter "A".

import com.styleconverter.runtime.PropertyTracker
import com.styleconverter.runtime.core.ir.IRComponent
import com.styleconverter.runtime.core.ir.IRLog
import com.styleconverter.runtime.core.ir.IRProperty
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull

/** Log tag for this bridge's named refusals — shared with the fold's. */
private const val TAG = "PseudoTextFold"

object PseudoTextBridge {

    /**
     * Declarations an inline text run consumes or may ignore without lying.
     * `content` is the run itself; the counter mutations (css-lists-3 §4)
     * are INPUTS to the producer's bake and are already resolved inside the
     * baked `_text`, so re-running them here could only disagree. Byte-
     * parallel with the iOS twin's `inertDeclarations`.
     */
    private val INERT_DECLARATIONS = setOf(
        "content", "counter-increment", "counter-reset", "counter-set"
    )

    /**
     * Box-model declarations that are SPEC-DEAD when the same bucket also
     * declares `display: contents` — css-display-3 §2.5: a box-less element
     * has nothing for border/background/size to apply to. Outside that
     * shape each of these means a real box and refuses the text fold.
     * Byte-parallel with the iOS twin's `boxUnderContents`.
     */
    private val BOX_UNDER_CONTENTS = setOf(
        "border", "background", "background-color", "width", "height"
    )

    /** The styling trio [PseudoStyledDeclarations] can type. */
    private val STYLING_TRIO = setOf("color", "font-size", "font-family")

    /** One foldable run: the baked glyphs plus its typed own styling. */
    data class InlineRun(val text: String, val styling: List<IRProperty>)

    /**
     * The bucket's foldable run, or null when there is nothing an inline
     * text run can honestly express (a declaration outside the consumable
     * set, an unconvertible style value, or no baked glyphs).
     */
    fun inlineRun(bucket: JsonObject, host: IRComponent): InlineRun? {
        // The raw CSS declarations map — extractor-owned and forwarded
        // verbatim (schema/spec/01-envelope.md). Absent map = a bare text
        // bucket: nothing to gate on.
        val decls = bucket["properties"] as? JsonObject
        // Pre-scan for `display: contents`, which flips the meaning of the
        // box set below and so must be known before the walk. The token
        // test is exact because `contents` is exclusive in the display
        // grammar (css-display-3 §2 — it never pairs like `block flow`).
        val displayContents = decls?.entries
            ?.firstOrNull { it.key.trim().lowercase() == "display" }
            ?.let { (it.value as? JsonPrimitive)?.contentOrNull?.trim()?.lowercase() == "contents" }
            ?: false
        // Does the host declare its own font-size? A RELATIVE pseudo size
        // resolves against the ORIGINATING ELEMENT's size (css-values-4
        // §6.1.1), but a FontSize appended to the host's own list resolves
        // against what the HOST inherited — a different base. Refuse that
        // combination rather than paint the wrong size (zero carriers in
        // the wave49-final corpus: contain-content-011's host declares only
        // `width`, so its `3em` resolves against the 16px default, right).
        val hostDeclaresFontSize = host.properties.any { it.type == "FontSize" }
        // Accumulators for the walk.
        var refused = false
        val styling = mutableListOf<IRProperty>()
        // Sorted so a multi-key refusal logs in a deterministic order.
        if (decls != null) for (prop in decls.keys.sorted()) {
            // css-syntax-3 §5: property names are ASCII case-insensitive.
            val key = prop.trim().lowercase()
            // Consumed by the run itself / already inside the bake.
            if (key in INERT_DECLARATIONS) continue
            // `display: contents` IS this fold's own semantics (the element
            // generates no box, only its content) — consumed. Any OTHER
            // display value still refuses below: a block/none pseudo is not
            // an inline text run.
            if (key == "display" && displayContents) continue
            // Box declarations are spec-dead under display:contents.
            if (key in BOX_UNDER_CONTENTS && displayContents) continue
            // The declaration's raw value; a non-primitive is not a
            // declarations map at all and refuses like any unknown shape.
            val raw = (decls[prop] as? JsonPrimitive)?.contentOrNull
            // The styling trio converts to the typed wire shapes the one
            // existing text pipeline already resolves.
            if (raw != null && key in STYLING_TRIO) {
                when (val conv = PseudoStyledDeclarations.convert(key, raw)) {
                    // Typed — collected for the caller's uniformity gate.
                    is PseudoStyledDeclarations.Conversion.Typed -> {
                        // The em/%/rem base guard above: only a relative
                        // FontSize is at risk, and only under a host that
                        // declares its own size.
                        if (key == "font-size" && hostDeclaresFontSize &&
                            !raw.trim().lowercase().endsWith("px")
                        ) {
                            PropertyTracker.markUnhandled("PseudoText::after/font-size")
                            IRLog.warn(TAG, "::after 'font-size: $raw' needs the host's own size as its em base — bucket not folded")
                            refused = true
                        } else {
                            styling += conv.property
                        }
                        continue
                    }
                    // A CSS-wide keyword collapsing to the run's existing
                    // default — consumed, correctly, as a no-op.
                    PseudoStyledDeclarations.Conversion.Inert -> continue
                    // Unparseable flavour (var()/calc()/keyword sizes) —
                    // falls through to the named refusal below, because
                    // folding the text with the WRONG ink/size would
                    // half-render the bucket.
                    PseudoStyledDeclarations.Conversion.Unsupported -> {}
                    // convert() returns null only for names outside the
                    // trio, which this branch has already excluded.
                    null -> {}
                }
            }
            // No silent fallthrough: name the refused declaration, then
            // keep scanning so EVERY offending declaration gets a line.
            PropertyTracker.markUnhandled("PseudoText::after/$key")
            IRLog.warn(TAG, "::after declaration '$prop' is beyond the inline-text fold — bucket not folded")
            refused = true
        }
        // Any refusal leaves the bucket on its existing wrapper path —
        // exactly the behaviour this platform had before the seam, now
        // named instead of silent.
        if (refused) return null
        // The baked literal. `_text` is what the WPT extractor emits today;
        // `text` is the v2 spelling — web tolerates both in exactly this
        // order (PseudoNodeRenderer's `p._text ?? p.text`), so mirror it.
        val text = (bucket["_text"] as? JsonPrimitive)?.takeIf { it.isString }?.content
            ?: (bucket["text"] as? JsonPrimitive)?.takeIf { it.isString }?.content
        // No bake at all: a `content` the producer could not resolve
        // (e.g. `attr()` on a subtree that never laid out). Web renders the
        // empty string for these too — name the gap, fold nothing.
        if (text == null) {
            if (decls?.keys?.any { it.trim().lowercase() == "content" } == true) {
                PropertyTracker.markUnhandled("PseudoText::after/unbaked")
                IRLog.warn(TAG, "::after content has no baked _text — rendered as empty")
            }
            return null
        }
        // `content: ""` bakes an EMPTY string: a legal pseudo that paints no
        // glyphs (css-content-3 §2), so there is nothing to fold.
        return if (text.isEmpty()) null else InlineRun(text, styling)
    }
}
